build/model_dist.py

```diff
@@ -0,0 +1,277 @@
+        self.register_buffer("freqs_cis", freqs_cis, persistent=True)
+        causal_mask = torch.tril(
+            torch.ones(self.max_seq_length, self.max_seq_length, dtype=torch.bool)
+        )
+        self.register_buffer("causal_mask", causal_mask, persistent=True)
+
+    def forward(self, idx: Tensor, input_pos: Optional[Tensor] = None) -> Tensor:
+        assert self.freqs_cis is not None, "Caches must be initialized first"
+        mask = self.causal_mask[None, None, input_pos]
+        freqs_cis = self.freqs_cis[input_pos]
+        x: DTensor = self.tok_embeddings(idx)
+        # TODO: sequence parallelize this
+
+        for _, layer in enumerate(self.layers):
+            x = layer(x, input_pos, freqs_cis, mask)
+        x = self.norm(x)
+        logits = self.output(x)
+        # print(f"logits shape: {logits.shape}")
+        return logits
+
+    @classmethod
+    def from_name(cls, name: str):
+        return cls(TransformerArgs.from_name(name))
+
+    @classmethod
+    def from_table(cls, name: str):
+        return cls(TransformerArgs.from_table(name))
+
+    @classmethod
+    def from_params(cls, params_path: str):
+        return cls(TransformerArgs.from_params(params_path))
+
+    @classmethod
+    def from_gguf(cls, gguf_path: str, **kwargs):
+        from build.gguf_loader import load_model_and_state_dict
+
+        model, state_dict = load_model_and_state_dict(gguf_path, **kwargs)
+        if state_dict != {}:
+            model.load_state_dict(state_dict, assign=True)
+        return model
+
+
+class TransformerBlock(nn.Module):
+    def __init__(self, config: TransformerArgs) -> None:
+        super().__init__()
+        self.attention = Attention(config)
+        self.feed_forward = FeedForward(config)
+        self.ffn_norm = RMSNorm(config.dim, config.norm_eps)
+        self.attention_norm = RMSNorm(config.dim, config.norm_eps)
+
+    def forward(
+        self, x: Tensor, input_pos: Tensor, freqs_cis: Tensor, mask: Tensor
+    ) -> Tensor:
+        h = x + self.attention(self.attention_norm(x), freqs_cis, mask, input_pos)
+        out = h + self.feed_forward(self.ffn_norm(h))
+        return out
+
+
+class Attention(nn.Module):
+    def __init__(self, config: TransformerArgs):
+        super().__init__()
+        assert config.dim % config.n_heads == 0
+
+        # key, query, value projections for all heads, but in a batch
+        # total_head_dim = (config.n_heads + 2 * config.n_local_heads) * config.head_dim
+        # self.wqkv = nn.Linear(config.dim, total_head_dim, bias=False)
+        wq = nn.Linear(config.dim, config.n_heads * config.head_dim, bias=False)
+        wk = nn.Linear(
+            config.dim, config.n_local_heads * config.head_dim, bias=False
+        )
+        wv = nn.Linear(
+            config.dim, config.n_local_heads * config.head_dim, bias=False
+        )
+        wo = nn.Linear(config.dim, config.dim, bias=False)
+
+        self.wq = parallelize_module(wq, device_mesh, Colwise)
```

Comment:
wondering if it's cleaner to just go ahead and merge the self.wx creation and the parallelize into single calls ala:
```
self.wq = parallelize_module(
                           nn.Linear(config.dim, config.n_heads * config.head_dim, bias=False,) 
                           device_mesh, Colwise
                           )
```